crates/sparrow-runtime/src/prepare/column_behavior.rs

```diff
@@ -0,0 +1,524 @@
+                let column = batch.column(*index);
+                error_stack::ensure!(
+                    *nullable || column.null_count() == 0,
+                    Error::NullInNonNullableColumn {
+                        field: batch.schema().field(*index).name().to_owned(),
+                        null_count: column.null_count()
+                    }
+                );
+                let previous_keys = metadata.previous_keys.borrow_mut();
+                // 1. Hash the current entity key column to a UInt64Array
+                let entity_column = sparrow_kernels::hash::hash(column)
+                    .into_report()
+                    .change_context(Error::PreparingColumn)?;
+                // 2. Convert the hash column specifically to a UInt64Array from a dyn array.
+                let indices: UInt64Array = entity_column
+                    .iter()
+                    .flatten()
+                    .enumerate()
+                    .filter_map(|(index, key_hash)| {
+                        if previous_keys.insert(key_hash) {
+                            Some(index as u64)
+                        } else {
+                            None
+                        }
+                    })
+                    .collect();
+                let new_hash_keys = arrow::compute::take(&entity_column, &indices, None)
+                    .into_report()
+                    .change_context(Error::PreparingColumn)?;
+                let new_keys = arrow::compute::take(column, &indices, None)
+                    .into_report()
+                    .change_context(Error::PreparingColumn)?;
+                metadata
+                    .add_entity_keys(new_hash_keys, new_keys)
+                    .into_report()
+                    .change_context(Error::PreparingColumn)?;
+                Arc::new(entity_column)
+            }
+            ColumnBehavior::ExecuteEntityKey { index, nullable } => {
+                // Hash the entity column and update the key hash table.
+                let key_hash_inverse = key_hash_inverse.expect("key_hash_inverse");
+                let keys = batch.column(*index);
+                error_stack::ensure!(
+                    *nullable || keys.null_count() == 0,
+                    Error::NullInNonNullableColumn {
+                        field: batch.schema().field(*index).name().to_owned(),
+                        null_count: keys.null_count()
+                    }
+                );
+                // 1. Hash the current entity key column to a UInt64Array
+                let key_hashes = sparrow_kernels::hash::hash(keys)
+                    .into_report()
+                    .change_context(Error::PreparingColumn)?;
+                // 2. Update key hash table
+                futures::executor::block_on(key_hash_inverse.add(keys.clone(), &key_hashes))
+                    .into_report()
+                    .change_context(Error::Internal)?;
```

Comment:
This was not easily avoidable, afaict. Any suggestions (even if we skip doing them for now)?

Reply:
I think it may be better to return the `keys` and `key_hashes` columns, and have this happen outside this method (possibly in the stream scan operation), as noted above. That way we don't need to do the block on, etc.

Reply:
Hm - I could update it in the `execute` method maybe..that's the only async context. Then I can give the `Scan` operation the the `ThreadSafeKeyHashInverse` and the `key_column_name`.